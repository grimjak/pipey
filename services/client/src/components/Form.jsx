import React from 'react';

const Form = (props) => {
    return (
        <div>
            <h1 className="title is-1">{props.formType}</h1>
            <hr/><br/>
            <form onSubmit={(event) => props.handleUserFormSubmit(event)}>
                <div className="field">
                    <input
                        name="username"
                        className="input is-medium"
                        type="text"
                        placeholder="Enter username"
                        required
                        value={props.formData.username}
                        onChange={props.handleFormChange}
                    />
                </div>
                <div className="field">
                    <input
                        name="password"
                        className="input is-medium"
                        type="password"
                        placeholder="Enter password"
                        required
                        value={props.formData.password}
                        onChange={props.handleFormChange}
                    />
                </div>
                <input
                    type="submit"
                    className="button is-primary is-medium is-fullwidth"
                    value="Submit"
                />
            </form>
        </div>
    )
};

export default Form;