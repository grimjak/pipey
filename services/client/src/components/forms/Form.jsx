import React, { Component } from 'react';
import axios from 'axios';
import { Redirect } from 'react-router-dom';
import { loginFormRules } from './form-rules.js';
import FormErrors from './FormErrors.jsx';

class Form extends Component {
    constructor (props) {
        super(props);
        this.state= {
            formData: {
                username: '',
                password: '',
            },
            loginFormRules: loginFormRules,
            valid: false,
        };
        this.handleUserFormSubmit = this.handleUserFormSubmit.bind(this);
        this.handleFormChange = this.handleFormChange.bind(this);
    };
    componentDidMount() {
        this.clearForm();
        this.validateForm();
    };
    componentWillReceiveProps(nextProps) {
        if (this.props.formType !== nextProps.formType) {
            this.clearForm();
            this.validateForm();
        };
    };
    clearForm() {
        this.setState({
            formData: {username: '', password: ''}
        });
    };
    handleFormChange(event){
        const obj = this.state.formData;
        obj[event.target.name] = event.target.value;
        this.setState(obj);
        this.validateForm();
    };
    handleUserFormSubmit(event){
        event.preventDefault();
        const formType = this.props.formType;
        const data = {
            username: this.state.formData.username,
            password: this.state.formData.password
        };
        const url = `${process.env.REACT_APP_USERS_SERVICE_URL}/auth/${formType}`
        axios.post(url, data)
        .then((res) => {
            this.clearForm();
            this.props.loginUser(res.data.auth_token);
        })
        .catch((err) => {
            console.log("error");
            this.props.createMessage('User does not exist.', 'danger'); 
        });
    };
    validateForm() {
        const self = this;
        const formData = this.state.formData;
        self.resetRules();
        //validate rules here
        if (self.props.formType === 'login') {
            const formRules = self.state.loginFormRules;
            if (formData.username.length > 0) formRules[0].valid = true;
            if (formData.password.length > 0) formRules[1].valid = true;
            self.setState({loginFormRules: formRules});
            if (self.allTrue()) self.setState({valid: true});
        }
    };
    allTrue() {
        let formRules = loginFormRules;
        for (const rule of formRules) {
            if (!rule.valid) return false;
        }
        return true;
    };
    resetRules() {
        const loginFormRules = this.state.loginFormRules;
        for (const rule of loginFormRules) {
            rule.valid = false;
        }
        this.setState({loginFormRules: loginFormRules});
        this.setState({valid: false})
    };
    render() {
        if(this.props.isAuthenticated) {
            return <Redirect to='/' />;
        };
        let formRules = this.state.loginFormRules;
        return (
            <div>
                <h1 className="title is-1">{this.props.formType}</h1>
                <hr/><br/>
                <FormErrors
                    formType={this.props.formType}
                    formRules={formRules}
                />
                <form onSubmit={(event) => this.handleUserFormSubmit(event)}>
                    <div className="field">
                        <input
                            name="username"
                            className="input is-medium"
                            type="text"
                            placeholder="Enter username"
                            required
                            value={this.state.formData.username}
                            onChange={this.handleFormChange}
                        />
                    </div>
                    <div className="field">
                        <input
                            name="password"
                            className="input is-medium"
                            type="password"
                            placeholder="Enter password"
                            required
                            value={this.state.formData.password}
                            onChange={this.handleFormChange}
                        />
                    </div>
                    <input
                        type="submit"
                        className="button btn-primary btn-lg btn-block"
                        value="Submit"
                        disabled={!this.state.valid}
                    />
                </form>
            </div>
        )
    };
};

export default Form;