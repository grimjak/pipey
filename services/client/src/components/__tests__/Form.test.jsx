import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Form from '../Form';

const testData = [
    {
        formType: 'Login',
        formData: {
            username: '',
            password: ''
        },
        handleUserFormSubmit: jest.fn(),
        handleFormChange: jest.fn(),
        isAuthenticated: false,
    },
];

describe('When not authenticated', () =>{
    testData.forEach((e1) => {
        const component = <Form {...e1} />;
        it(`${e1.formType} Form renders properly`, () => {
            const wrapper = shallow(component);
            const h1 = wrapper.find('h1');
            expect(h1.length).toBe(1);
            expect(h1.get(0).props.children).toBe(e1.formType);
            const formGroup = wrapper.find('.field');
            expect(formGroup.length).toBe(Object.keys(e1.formData).length);
            expect(formGroup.get(0).props.children.props.name).toBe(Object.keys(e1.formData)[0]);
            expect(formGroup.get(0).props.children.props.value).toBe('');
        });
        it(`${e1.formType} Form submits the form properly`, () => {
            const wrapper = shallow(component);
            const input = wrapper.find('input[type="text"]');
            expect(e1.handleUserFormSubmit).toHaveBeenCalledTimes(0);
            expect(e1.handleFormChange).toHaveBeenCalledTimes(0);
            input.simulate('change', { target: {value: 'tb'}});
            expect(e1.handleFormChange).toHaveBeenCalledTimes(1);
            wrapper.find('form').simulate('submit', e1.formData);
            expect(e1.handleUserFormSubmit).toHaveBeenCalledWith(e1.formData);
            expect(e1.handleUserFormSubmit).toHaveBeenCalledTimes(1);
        });
        it(`{e1.formType} Form renders a snapshot properly`, () => {
            const tree = renderer.create(component).toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});

describe('When authenticated', () => {
    testData.forEach((e1) => {
        const component = <Form
            formType={e1.formType}
            formData={e1.formData}
            isAuthenticated={true}
        />;
        it(`$e1.formType} redirects properly`, () => {
            const wrapper = shallow(component);
            expect(wrapper.find('Redirect')).toHaveLength(1);
        });
    });
});
