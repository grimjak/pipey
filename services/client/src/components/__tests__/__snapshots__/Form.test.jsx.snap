// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When not authenticated {e1.formType} Form renders a snapshot properly 1`] = `
<div>
  <h1
    className="title is-1"
  >
    Login
  </h1>
  <hr />
  <br />
  <div>
    <ul
      className="validation-list"
    >
      <li
        className="error"
      >
        Username is required.
      </li>
      <li
        className="error"
      >
        Password is required.
      </li>
    </ul>
    <br />
  </div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <input
        className="input is-medium"
        name="username"
        onChange={[Function]}
        placeholder="Enter username"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="field"
    >
      <input
        className="input is-medium"
        name="password"
        onChange={[Function]}
        placeholder="Enter password"
        required={true}
        type="password"
        value=""
      />
    </div>
    <input
      className="button btn-primary btn-lg btn-block"
      disabled={true}
      type="submit"
      value="Submit"
    />
  </form>
</div>
`;
